File: flightcontext.go
package agentsdk

import "sync"

// FlightContext is a goroutine-safe key-value store for recording in-flight
// state that should be captured on crash (signal or panic).
//
// It uses sync.RWMutex to favor read-heavy workloads where agents frequently
// read context while only occasionally updating it. All methods are safe for
// concurrent use.
type FlightContext struct {
	mu   sync.RWMutex
	data map[string]interface{}
}

// NewFlightContext creates an empty FlightContext.
func NewFlightContext() *FlightContext {
	return &FlightContext{
		data: make(map[string]interface{}),
	}
}

// Get retrieves the value for the given key. Returns nil if the key does not exist.
func (fc *FlightContext) Get(key string) interface{} {
	fc.mu.RLock()
	defer fc.mu.RUnlock()
	return fc.data[key]
}

// Set stores a value for the given key, overwriting any existing value.
func (fc *FlightContext) Set(key string, value interface{}) {
	fc.mu.Lock()
	defer fc.mu.Unlock()
	fc.data[key] = value
}

// Delete removes the value for the given key. No-op if the key does not exist.
func (fc *FlightContext) Delete(key string) {
	fc.mu.Lock()
	defer fc.mu.Unlock()
	delete(fc.data, key)
}

// Snapshot returns a shallow copy of the current key-value map.
// The returned map is safe to mutate without affecting the FlightContext.
// Snapshot does NOT deep-copy interface{} values — callers storing mutable
// references (slices, maps, pointers) should be aware of this.
func (fc *FlightContext) Snapshot() map[string]interface{} {
	fc.mu.RLock()
	defer fc.mu.RUnlock()
	snapshot := make(map[string]interface{}, len(fc.data))
	for k, v := range fc.data {
		snapshot[k] = v
	}
	return snapshot
}

File: sdks/python/agentsdk/flightcontext.py
"""FlightContext — thread-safe key-value store for in-flight state.

Records in-flight state that should be captured on crash (signal or panic).
Uses ``threading.RLock`` for concurrent access safety, favoring read-heavy
workloads where agents frequently read context while only occasionally
updating it.

All methods are safe for concurrent use.
"""

from __future__ import annotations

import threading
from typing import Any, Dict, Optional


class FlightContext:
    """Thread-safe key-value store for recording in-flight state."""

    def __init__(self) -> None:
        self._mu: threading.RLock = threading.RLock()
        self._data: Dict[str, Any] = {}

    # -- Read operations -----------------------------------------------------

    def get(self, key: str) -> Optional[Any]:
        """Retrieve the value for *key*. Returns ``None`` if the key does not exist."""
        with self._mu:
            return self._data.get(key)

    def snapshot(self) -> Dict[str, Any]:
        """Return a shallow copy of the current key-value map.

        The returned dict is safe to mutate without affecting the
        FlightContext.  Snapshot does **not** deep-copy values — callers
        storing mutable references (lists, dicts) should be aware of this.
        """
        with self._mu:
            return dict(self._data)

    # -- Write operations ----------------------------------------------------

    def set(self, key: str, value: Any) -> None:
        """Store *value* for *key*, overwriting any existing value."""
        with self._mu:
            self._data[key] = value

    def delete(self, key: str) -> None:
        """Remove the value for *key*. No-op if the key does not exist."""
        with self._mu:
            self._data.pop(key, None)
